### PR TITLE
KAFKA-5540 : Deprecate and remove internal converter configs

### PR DESCRIPTION
**Pull request by umesh9794**
This PR deprecates the following configs: 

`internal.key.converter`
`internal.value.converter`

And removes their usage from connect-*.properties files.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by umesh9794** [comment]
@ewencp @rhauch please review the PR and let me know your thoughts.

**Comment by rhauch** [comment]
@umesh9794, thanks for the PR. The `internal.key.converter` and `internal.value.converter` don't currently have defaults, so in addition to deprecating these configuration properties don't we also want to set the default to `org.apache.kafka.connect.json.JsonConverter` *and* disable their use of schemas? Changing the default is pretty easy, but setting the converter-specific properties to disable schemas will require changes in the `Worker` constructor (where the converters are currently instantiated) to set the properties if the JSON converter is used and the `internal.key.converter.schemas.enable` and `internal.key.converter.schemas.enable` are not explicitly set.

**Comment by umesh9794** [comment]
Thanks @rhauch for the review comments. I updated the PR to not use `internal.key.converter.schemas.enable` and `internal.value.converter.schemas.enable` in the `Worker` constructor. Instead I used `key.converter` and `value.converter` as `JsonConverter` still needed to be configured. It took me some time to read the `WorkerTest` which is why this PR got delayed. Please review the updated PR and let me know if I am missing something.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java [hunk 1]
```diff
@@ -61,24 +61,26 @@
...
     public static final String INTERNAL_VALUE_CONVERTER_CLASS_CONFIG = "internal.value.converter";
     public static final String INTERNAL_VALUE_CONVERTER_CLASS_DOC =
             "Converter class used to convert between Kafka Connect format and the serialized form that is written to Kafka." +
                     " This controls the format of the values in messages written to or read from Kafka, and since this is" +
                     " independent of connectors it allows any connector to work with any serialization format." +
                     " Examples of common formats include JSON and Avro." +
                     " This setting controls the format used for internal bookkeeping data used by the framework, such as" +
-                    " configs and offsets, so users can typically use any functioning Converter implementation.";
+                    " configs and offsets, so users can typically use any functioning Converter implementation." +
+                    " This config is deprecated, refer KAFKA-5540";
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: I'd suggest the last sentence should be something like "The JSON converter should always be used with schemas disabled, and this is now the default. This property will be removed in a future version."

**Comment by rhauch** [comment]
@umesh9794, I'm concerned about the latest commit. Remember that *per the KIP* in this release we are only deprecating these configuration properties and using defaults so that users will not have to specify them anymore. However, users should still be able to use the configuration properties, and we are **not** removing the effectiveness of those properties until the next major version (2.0). The last commit effectively does all of this at once.

Although unlikely, it is possible that some users have configured the internal key and value converters to something other than the JSON converter. If we were to make the changes in this PR, those users would not be able to read in any messages from the internal topics. We need to give those users time to migrate to the JSON converter.

Therefore, in this PR we need to still use the `internal.key.converter` and `internal.value.converter` properties but need to use `org.apache.kafka.connect.json.JsonConverter` as the default value for both. This will allow users to start removing those properties, but if they supply something different than the default it will still work for the remaining 1.x releases.

However, our configuration files always configured the JsonConverter to *disable the use of schemas*, but this is not the default for the JsonConverter. This will take a bit more effort to change in the short term: the `Worker` class instantiates and configures the internal converters, but before this we need to check whether the `internal.value.converter.enable.schemas` and `internal.key.converter.enable.schemas` properties were set. If they were, we should use them; if not, we should set them to `false` (which is not the default!).

I hope this makes sense.

**Review by rhauch — CHANGES_REQUESTED** [review]
In addition to the previous comment about how this PR does too much by removing the effectiveness of these properties, the changes to the documentation for the configuration properties could use some tweaking as well.

**Comment by umesh9794** [comment]
Thanks @rhauch for the detailed review comments. I've applied the comments in the related classes, please have a look and let me know your thoughts.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java (hunk 1 above)
Comment (umesh9794) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: Agreed and used the mentioned lines in the config doc. Please have a look.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 2]
```diff
@@ -100,20 +102,39 @@ public Worker(
...
-                config.getClass(WorkerConfig.INTERNAL_KEY_CONVERTER_CLASS_CONFIG).getName(),
-                config
-        );
+        // Internal key/value converters are currently deprecated, instead JsonConverter is used by default, refer KIP-174.
+        // Currently Internal converters are required properties, thus getClass won't return null.
+
+        boolean isJsonConverterKeyClassUsed = config.getClass(WorkerConfig.INTERNAL_KEY_CONVERTER_CLASS_CONFIG).getName().equals("org.apache.kafka.connect.json.JsonConverter");
+        boolean isJsonConverterValueClassUsed = config.getClass(WorkerConfig.INTERNAL_VALUE_CONVERTER_CLASS_CONFIG).getName().equals("org.apache.kafka.connect.json.JsonConverter");
+
+        if (!isJsonConverterKeyClassUsed && !isJsonConverterValueClassUsed) {
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:111]: Is there a reason not to handle each case independently? I'm not clear on the reasoning for checking both of these.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 3]
```diff
@@ -100,20 +102,39 @@ public Worker(
...
+            this.internalKeyConverter =
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:112]: Doesn't the KIP say we're going to log a warning if the setting is specified? I don't see that anywhere, just the configs marked deprecated.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java [hunk 4]
```diff
@@ -61,24 +61,28 @@
                     " This controls the format of the values in messages written to or read from Kafka, and since this is" +
                     " independent of connectors it allows any connector to work with any serialization format." +
                     " Examples of common formats include JSON and Avro.";
-
+    @Deprecated
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:64]: These doc changes aren't sufficient due to the way Kafka configs work. Without a default provided, parsing the configs will throw an exception (before you ever even hit the code you have in `Worker.java`). You'd need to change the default value to actually be `JsonConverter`. (This is actually a bit limiting once you get to the parsed version of the config because you can't tell if the user actually *specified `JsonConverter` explicitly or is using the default; however, I think this is a limit of the `CLASS` type in `ConfigDef`. If it was just a `STRING` we could leave it empty by default, detect that and fall back to `JsonConverter`, and know based on this whether we should be using the user's settings for each converter or just specify our set of defaults.)

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 5]
```diff
@@ -100,20 +102,39 @@ public Worker(
...
+                    plugins.newConverter(config.getClass(WorkerConfig.INTERNAL_KEY_CONVERTER_CLASS_CONFIG).getName(), config);
+            this.internalValueConverter =
+                    plugins.newConverter(config.getClass(WorkerConfig.INTERNAL_VALUE_CONVERTER_CLASS_CONFIG).getName(), config);
+        } else {
+            this.internalKeyConverter = plugins.newConverter(JsonConverter.class.getName(), config);
+            this.internalValueConverter = plugins.newConverter(JsonConverter.class.getName(), config);
+        }
+
+        boolean isInternalConverterSchemaConfigured = config.originals().containsKey("internal.key.converter.schemas.enable") && config.originals().containsKey("internal.value.converter.schemas.enable");
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:121]: This is interesting and something I don't think we really considered with compatibility. Technically the default values only apply to the values in the ConfigDef, which means the actual default for `schemas.enable` is the default in `JsonConverter`, which is `true`, not the `false` value we setup in the example configs.

In practice I think making this change won't matter since I think offsets don't work with `JsonConverter` and `schemas.enable=true`, but it would be good to verify (and maybe even update the KIP compatibility discussion to clarify). There's a JIRA related to that issue which would be good justification for why we are pretty sure nobody is using it in that mode, I think it was closed due to this KIP but should be findable easily in JIRA search.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 6]
```diff
@@ -100,20 +102,39 @@ public Worker(
...
+
+        Map<String, String> userConfigs = config.originalsStrings();
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:123]: I'm not sure the handling of configs here is quite right. I think it would make more sense to check if there are *any* properties set with the prefix `internal.key.converter` (and `internal.value.converter` for that case). There are other configs that could be overridden and really at this point, if the user has specified *anything* explicitly, we want to still respect those settings. Only after the deprecation period should there be cases where we might ignore/modify settings they explicitly specify.

In other words, I'm not sure a version of this code that does not maintain use of `config.originalsWithPrefix("internal.key.converter")` and `config.originalsWithPrefix("internal.value.converter")` can properly maintain compatibility.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 2 above)
Comment (umesh9794) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:111]: I missed the fact that both `internal.key.converter` and `internal.value.converter` can be configured with different converter classes, hence I checked both whether they are configured with `JsonConverter` class or not. I will split this check for both of the configurations.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 3 above)
Comment (umesh9794) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:112]: Ah, sorry I missed to log that. I'll log the warning if these configurations value is set by user explicitly.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 6 above)
Comment (umesh9794) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:123]: Agree, however I am not pretty sure if we have more configs other than `internal.key.converter.schemas.enable` and `internal.value.converter.schemas.enable` which qualify the prefix search using `config.originalsWithPrefix("internal.key.converter.")` and `config.originalsWithPrefix("internal.value.converter.")`. That is why I used the full name of these configs instead of filtering them with prefix. 

Is there any existing config which has `internal.key|value.converter.` as prefix and we need to take care of that in Worker's context ?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java (hunk 4 above)
Comment (umesh9794) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:64]: Yes, I can understand that when we don't provide any defaults to the configs while defining the `ConfigDef`, it is considered as required and one has to configure that config explicitly. I believe by configuring the default as `JsonConverter`, we can still check whether some converter other than `JsonConverter` is provided by user. Although it will certainly be difficult to distinguish where this `JsonConverter` is set (its getting populated as default or provided by user), still the code will proceed further to follow the KIP and to use the user provided converters. I think it is the expected behavior as per KIP and setting the default for these deprecated configs should not break the existing users's code.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 5 above)
Comment (umesh9794) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:121]: Yes, I believe you are referring [KAFKA-3988](https://issues.apache.org/jira/browse/KAFKA-3988). Also we have already mentioned this in our KIP, so are you suggesting to edit the KIP to add something like: "KAFKA-3988 helped us to track the use of Converters with schema. Since we don't find any such usage so we assume that nobody is using Converters with schema." 

Let me know if I missed anything.

**Comment by kkonstantine** [comment]
I'm closing this PR at the moment because [KAFKA-5540](https://issues.apache.org/jira/browse/KAFKA-5540) ended up being discussed with [KIP-174](https://cwiki.apache.org/confluence/display/KAFKA/KIP-174+-+Deprecate+and+remove+internal+converter+configs+in+WorkerConfig) which has now been adopted and implemented (PR here: https://github.com/apache/kafka/pull/4693) 

Thanks for highlighting this improvement @umesh9794 in the first place! I believe it required a KIP as well as the release of a major version (AK 2.0.0) to be introduced with.
